pkg/util/kubernetes/kubernetes.go

```diff
@@ -598,13 +598,21 @@ func NewNatsPodSpec(name, clusterName string, cs v1alpha2.ClusterSpec, owner met
 		container = containerWithRequirements(container, cs.Pod.Resources)
 	}
 
+	// Grab the A record that will correspond to the current pod so we can use it as the client and cluster advertise host.
+	// This helps with client failover, and with avoiding route connection errors in TLS-enabled clusters.
+	advertiseHost := fmt.Sprintf("%s.%s.%s.svc", name, ManagementServiceName(clusterName), namespace)
+
 	// Rely on the shared configuration map for configuring the cluster.
 	cmd := []string{
 		constants.NatsBinaryPath,
 		"-c",
 		constants.ConfigFilePath,
 		"-P",
 		constants.PidFilePath,
+		"--cluster_advertise",
+		advertiseHost,
+		"--client_advertise",
+		advertiseHost,
```

Comment:
cluster advertise is correct but client advertise won't be right in some cases (for example when setting the client ip to be the one from the kubelet node, though will work in some cases when the traffic is all internal)